### PR TITLE
Enable enrollments empty check in GenesisBlock

### LINKED ISSUE #1361: Agora accepts genesis blocks with no enrollments
Originally found in: https://github.com/bpfkorea/infrastructure/issues/89 (private)

We accidentally deployed an old config to our server, which meant the `testing` field in the `Config` was set to `false`:

https://github.com/bpfkorea/agora/blob/c0f000595de29fbe8ce837f7e0eb8a88e3d10c10/source/agora/common/Config.d#L115

When faucet was deployed, the nodes accepted the transactions. That's because currently, CoinNet and TestNet have the exact same genesis funding transactions:

https://github.com/bpfkorea/agora/blob/c0f000595de29fbe8ce837f7e0eb8a88e3d10c10/source/agora/consensus/data/genesis/Test.d#L135-L145

and:

https://github.com/bpfkorea/agora/blob/c0f000595de29fbe8ce837f7e0eb8a88e3d10c10/source/agora/consensus/data/genesis/Coinnet.d#L52

**However**, the CoinNet genesis block doesn't have **any** enrollments.

So the question is, why did the nodes load the CoinNet genesis block and accepted it, even though it has no enrollments in it? This is the reason:

https://github.com/bpfkorea/agora/blob/c0f000595de29fbe8ce837f7e0eb8a88e3d10c10/source/agora/consensus/validation/Block.d#L218-L220

In fact I made a comment about it when the PR was made:

https://github.com/bpfkorea/agora/pull/884#discussion_r437218299

### PR DESCRIPTION
**Pull request by TrustHenry**
Genesis blocks must contain enrollments.
Genesis blocks without enrollments cannot create the next block.

Related to #1361, #1411

### CODE REVIEW

**Comment by TrustHenry** [comment]
@Geod24 The first block start policy is required.
https://github.com/bpfkorea/agora/issues/1361#issuecomment-734531195

File: source/agora/consensus/validation/Block.d [hunk 1]
```diff
@@ -223,7 +223,7 @@ public string isGenesisBlockInvalidReason (const ref Block block) nothrow @safe
 
     // If there are no enrollments, return them here early.
     if (block.header.enrollments.length == 0)
-        return null;
+        return "GenesisBlock: Does not include enrollments";
```
Comment (linked0) [inline source/agora/consensus/validation/Block.d]: How about expressing in the same format as other messages, like "No enrollment in the block"

**Review by linked0 — COMMENTED** [review]
Good start!
One request and there's an error about Genesis block having no enrollment on the CI build. Would you please check it?

File: source/agora/consensus/validation/Block.d (hunk 1 above)
Comment (TrustHenry) [reply source/agora/consensus/validation/Block.d]: It's good, the sentence.

**Comment by TrustHenry** [comment]
> One request and there's an error about Genesis block having no enrollment on the CI build. Would you please check it?

It's written on the top.
This has been blocked due to the following issues.
#1361 (comment)

**Comment by AndrejMitrovic** [comment]
You can just add a temporary fake enrollment to the coinnet genesis block.

The real enrollments and transactions will be defined later.

**Comment by TrustHenry** [comment]
Ready for review.

**Comment by TrustHenry** [comment]
> You can just add a temporary fake enrollment to the coinnet genesis block.

@AndrejMitrovic Yes, I did temporarily defined enrollments and reconstructed coinnet genesis block.

File: source/agora/consensus/validation/Block.d [hunk 2]
```diff
@@ -223,7 +223,7 @@ public string isGenesisBlockInvalidReason (const ref Block block) nothrow @safe
 
     // If there are no enrollments, return them here early.
```
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Block.d]: Outdated. You can just remove this comment.

File: source/agora/consensus/validation/Block.d [hunk 3]
```diff
@@ -314,21 +321,22 @@ unittest
     {
         // Txs length check
         block.txs = null;
-        block.header.enrollments = null;
         assert(!block.isGenesisBlockValid());
 
         // at least 1 tx needed (todo: relax this?)
-        block.txs ~= makeNewTx();
+        block.txs = null;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Block.d]: It's already `null`, can you remove this line?

File: source/agora/consensus/validation/Block.d [hunk 4]
```diff
@@ -223,7 +223,7 @@ public string isGenesisBlockInvalidReason (const ref Block block) nothrow @safe
 
     // If there are no enrollments, return them here early.
     if (block.header.enrollments.length == 0)
-        return null;
+        return "GenesisBlock: No enrollment in the block";
```
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Block.d]:
```suggestion
        return "GenesisBlock: No enrollments in the block";
```

More proper English. :)

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]
Just a few comments, almost ready to go.

**Comment by TrustHenry** [comment]
@AndrejMitrovic Thanks for the reviews. All fixed it.

**Review by AndrejMitrovic — APPROVED** [review]

**Comment by AndrejMitrovic** [comment]
I wish you kept it the same layout like the test genesis. In the test genesis the first tx is the freeze, in yours its the pay.

It's ok, I'll fix it myself.

**Comment by TrustHenry** [comment]
> I wish you kept it the same layout like the test genesis. In the test genesis the first tx is the freeze, in yours its the pay.
> 
> It's ok, I'll fix it myself.

The order of transactions has changed because of the transaction hash sorting.
Is it weird?

**Comment by AndrejMitrovic** [comment]
No it's ok, it was just a little harder to create the enrollments. It's ok, thank you.

**Comment by TrustHenry** [comment]
> No it's ok, it was just a little harder to create the enrollments. It's ok, thank you.

Yes, it was hard to make the enrolments. :)

**Pull request merged**